diff --git a/top-level.nix b/top-level.nix
index b8662e9929..e1b06a8490 100644
--- a/top-level.nix
+++ b/top-level.nix
@@ -86,7 +86,7 @@
     in
     if replacements == [ ] then
       # Avoid IFD if possible, by sidestepping replaceDependencies if no replacements are specified.
-      baseSystemAssertWarn
+      baseSystem
     else
       (pkgs.replaceDependencies.override {
         replaceDirectDependencies = pkgs.replaceDirectDependencies.override {
@@ -94,7 +94,7 @@
         };
       })
         {
-          drv = baseSystemAssertWarn;
+          drv = baseSystem;
           inherit replacements cutoffPackages;
         };
 
@@ -109,7 +109,6 @@
 
 {
   imports = [
-    ../build.nix
     (mkRemovedOptionModule [
       "nesting"
       "clone"
@@ -342,7 +341,7 @@
 
     system.extraSystemBuilderCmds =
       optionalString config.system.copySystemConfiguration ''
-        ln -s '${import ../../../lib/from-env.nix "NIXOS_CONFIG" <nixos-config>}' \
+        ln -s '${import "${pkgs.path}/nixos/lib/from-env.nix" "NIXOS_CONFIG" <nixos-config>}' \
           "$out/configuration.nix"
       ''
       + optionalString (config.system.forbiddenDependenciesRegexes != [ ]) (
