diff --git a/modules.nix b/modules.nix
index 3330579952..bd93973164 100644
--- a/modules.nix
+++ b/modules.nix
@@ -616,7 +616,10 @@
           "_file"
           "key"
           "disabledModules"
+          "inputs"
           "imports"
+          "osImports"
+          "hmImports"
           "options"
           "config"
           "meta"
@@ -631,7 +634,10 @@
           _class = m._class or null;
           key = toString m.key or key;
           disabledModules = m.disabledModules or [ ];
+          inputs = m.inputs or { };
           imports = m.imports or [ ];
+          osImports = m.osImports or [ ];
+          hmImports = m.hmImports or [ ];
           options = m.options or { };
           config = addFreeformType (addMeta (m.config or { }));
         }
@@ -642,7 +648,10 @@
         _class = m._class or null;
         key = toString m.key or key;
         disabledModules = m.disabledModules or [ ];
+        inputs = m.inputs or { };
         imports = m.require or [ ] ++ m.imports or [ ];
+        osImports = m.osImports or [ ];
+        hmImports = m.hmImports or [ ];
         options = { };
         config = addFreeformType (
           removeAttrs m [
@@ -650,8 +659,11 @@
             "_file"
             "key"
             "disabledModules"
+            "inputs"
             "require"
             "imports"
+            "osImports"
+            "hmImports"
             "freeformType"
           ]
         );
@@ -2027,7 +2039,6 @@
           pushDownProperties
           unifyModuleSyntax
           ;
-        collectModules = collectModules null;
       };
 
   /**
@@ -2161,6 +2172,7 @@
   #       are just needed by types.nix, but are not meant to be consumed
   #       externally.
   inherit
+    collectModules
     defaultOrderPriority
     defaultOverridePriority
     defaultPriority
